pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/StagePlanner.java

```diff
@@ -160,18 +167,18 @@ private boolean canSkipShuffle(StageNode stageNode, KeySelector<Object[], Object
     Set<Integer> originSet = stageNode.getPartitionKeys();
     if (!originSet.isEmpty() && keySelector != null) {
       Set<Integer> targetSet = new HashSet<>(((FieldSelectionKeySelector) keySelector).getColumnIndices());
-      return targetSet.containsAll(originSet);
+      return originSet.containsAll(targetSet);
```

Comment:
This is a hack. This sets the distribution-type to SINGLETON for the join stage when the child stages are for different nodes.